subsys/net/ip/net_pkt.c

```diff
@@ -2176,6 +2188,472 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
+		current = new;
+		if (current->size > size) {
+			current->size = size;
+			size = 0;
+		} else {
+			size -= current->size;
+		}
+
+		if (timeout != K_NO_WAIT && timeout != K_FOREVER) {
+			u32_t diff = k_uptime_get_32() - alloc_start;
+
+			timeout -= min(timeout, diff);
+		}
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+		NET_FRAG_CHECK_IF_NOT_IN_USE(new, new->ref + 1);
+
+		net_pkt_alloc_add(new, false, caller, line);
+
+		NET_DBG("%s (%s) [%d] frag %p ref %d (%s():%d)",
+			pool2str(pool), get_name(pool), get_frees(pool),
+			new, new->ref, caller, line);
+#endif
+	}
+
+	return first;
+error:
+	if (first) {
+		net_buf_unref(first);
+	}
+
+	return NULL;
+}
+
+#else /* !CONFIG_NET_BUF_FIXED_DATA_SIZE */
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+static struct net_buf *pkt_alloc_buffer(struct net_buf_pool *pool,
+					size_t size, s32_t timeout,
+					const char *caller, int line)
+#else
+static struct net_buf *pkt_alloc_buffer(struct net_buf_pool *pool,
+					size_t size, s32_t timeout)
+#endif
+{
+	struct net_buf *buf;
+
+	buf = net_buf_alloc_len(pool, size, timeout);
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+	NET_FRAG_CHECK_IF_NOT_IN_USE(buf, buf->ref + 1);
+
+	net_pkt_alloc_add(buf, false, caller, line);
+
+	NET_DBG("%s (%s) [%d] frag %p ref %d (%s():%d)",
+		pool2str(pool), get_name(pool), get_frees(pool),
+		buf, buf->ref, caller, line);
+#endif
+
+	return buf;
+}
+
+#endif /* CONFIG_NET_BUF_FIXED_DATA_SIZE */
+
+static size_t pkt_buffer_length(struct net_pkt *pkt,
+				size_t size,
+				enum net_ip_protocol proto,
+				size_t existing)
+{
+	size_t max_len = net_if_get_mtu(net_pkt_iface(pkt));
+	sa_family_t family = net_pkt_family(pkt);
+
+	/* Family vs iface MTU */
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
+		max_len = max(max_len, NET_IPV6_MTU);
+	} else if (IS_ENABLED(CONFIG_NET_IPV4) && family == AF_INET) {
+		max_len = max(max_len, NET_IPV4_MTU);
+	}
+
+	max_len -= existing;
+	max_len = min(size, max_len);
+
+	return min(size, max_len);
+}
+
+static size_t pkt_headers_buffer_length(struct net_pkt *pkt,
+					sa_family_t family,
+					enum net_ip_protocol proto)
+{
+	size_t hdr_len = 0;
+
+	if (family == AF_UNSPEC) {
+		return  0;
+	}
+
+	/* Family header */
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
+		hdr_len += NET_IPV6H_LEN;
+	} else if (IS_ENABLED(CONFIG_NET_IPV4) && family == AF_INET) {
+		hdr_len += NET_IPV4H_LEN;
+	}
+
+	/* + protocol header */
+	if (IS_ENABLED(CONFIG_NET_TCP) && proto == IPPROTO_TCP) {
+		hdr_len += NET_TCPH_LEN + NET_TCP_MAX_OPT_SIZE;
+	} else if (IS_ENABLED(CONFIG_NET_UDP) && proto == IPPROTO_UDP) {
+		hdr_len += NET_UDPH_LEN;
+
+		if (IS_ENABLED(CONFIG_NET_RPL_INSERT_HBH_OPTION)) {
+			hdr_len += NET_RPL_HOP_BY_HOP_LEN;
+		}
+	} else if (proto == IPPROTO_ICMP || proto == IPPROTO_ICMPV6) {
+		hdr_len += NET_ICMPH_LEN;
+	}
+
+	NET_DBG("HDRs length estimation %zu", hdr_len);
+
+	return hdr_len;
+}
+
+static size_t pkt_get_size(struct net_pkt *pkt)
+{
+	struct net_buf *buf = pkt->buffer;
```

Comment:
I think this member should be called buffer**s** after all.

Reply:
no, because it represents the overall packet buffer. Fact that it might be made of one or multiple net_buf (in a list) is not relevant for the name here.

Reply:
> Fact that it might be made of one or multiple net_buf (in a list) is not relevant for the name here.

For us humans, it is. It wouldn't be relevant in a language with proper abstraction layer, like C++. But we have very thin, flat API here, where layout of data structures is part of API and thus naming should reflect that. Many people who just start with Zephyr IP stack would get an impression that there's a single buffer behind it and won't try to iterate on it. (Maybe your aim is to provide API which will discourage direct access to data structures, instead pushing access via abstracted method calls, but that's not how we had it (nor extra abstraction needed here, having thin abstractions for efficient code isn't bad)).